llvm/lib/Transforms/IPO/FunctionSpecialization.cpp

```diff
@@ -875,48 +909,67 @@ bool FunctionSpecializer::findSpecializations(Function *F, unsigned FuncSize,
       AllSpecs[Index].CallSites.push_back(&CS);
     } else {
       // Calculate the specialisation gain.
-      Bonus B;
+      Cost CodeSize;
       unsigned Score = 0;
       InstCostVisitor Visitor = getInstCostVisitorFor(F);
       for (ArgInfo &A : S.Args) {
-        B += Visitor.getSpecializationBonus(A.Formal, A.Actual);
+        CodeSize += Visitor.getCodeSizeSavingsForArg(A.Formal, A.Actual);
         Score += getInliningBonus(A.Formal, A.Actual);
       }
-      B += Visitor.getBonusFromPendingPHIs();
-
+      CodeSize += Visitor.getCodeSizeSavingsFromPendingPHIs();
 
-      LLVM_DEBUG(dbgs() << "FnSpecialization: Specialization bonus {CodeSize = "
-                        << B.CodeSize << ", Latency = " << B.Latency
-                        << ", Inlining = " << Score << "}\n");
-
-      FunctionGrowth[F] += FuncSize - B.CodeSize;
-
-      auto IsProfitable = [](Bonus &B, unsigned Score, unsigned FuncSize,
-                             unsigned FuncGrowth) -> bool {
+      auto IsProfitable = [&]() -> bool {
         // No check required.
         if (ForceSpecialization)
           return true;
+
+        unsigned CodeSizeSavings = getCostValue(CodeSize);
+        // TODO: We should only accumulate codesize increase of specializations
+        // that are actually created.
+        FunctionGrowth[F] += FuncSize - CodeSizeSavings;
```

Comment:
It is true indeed. We only keep `std::min(NumCandidates * MaxClones, unsigned(AllSpecs.size()))` per invocation of `FunctionSpecializer::run()`. It's not trivial to know upfront whether a specialization will be selected even if this lambda says it's profitable. However, if we move the FunctionGrowth accumulation just before the lambda returns true, does it not improve the likelihood of being created? (at least we know that MinCodeSizeSavings and MinLatencySavings requirements are satisfied at that point)

Reply:
Hmm, but we sometimes early exit due to inlining bonus. Bummer

Reply:
Perhaps we can move it here https://github.com/llvm/llvm-project/blob/56f75b5c1510cea7648cad0cb32e4e0810edd0d9/llvm/lib/Transforms/IPO/FunctionSpecialization.cpp#L735. At that point we know F, we know FuncSize and we know CodeSizeSavings (it is S.Score as long as we remove `Score += std::max(CodeSizeSavings, LatencySavings)`). Can you try that?

Reply:
> Hmm, but we sometimes early exit due to inlining bonus. Bummer

Yeah, it's somewhat annoying - I've put it where it is for this reason.

>Perhaps we can move it here ...

I think that this would be the ideal place to do the accumulation to `FunctionGrowth[F]`.

Could you clarify what you mean by the following:
> (it is S.Score as long as we remove Score += std::max(CodeSizeSavings, LatencySavings)). 

The Score is initially the Inlining Bonus, and then we add `std::max(CodeSizeSavings, LatencySavings))` - do you mean to just simply set `Score = CodeSizeSavings` in the case where we reach this point i.e. we are specializing due to CodeSizeSavings/LatencySavings rather than the Inlining Bonus?

Reply:
Yeah actually - that still wouldn't work because we wouldn't be able to distinguish if the Score represented CodeSizeSavings or inlining bonus. I think we'd need to add something like `CodeSizeCost` to the `Spec` struct in order to pass this information through to where we actually do the specializations?

My feeling is that it would make the most sense to leave this as it is for this PR and make these changes in a follow up PR, if you agree?

Reply:
Fair enough, it alredy an improvement and the problem you spotted is not introduced by this patch. Good we have a TODO.